File: windows/ClipSync.Core/Update/GitHubReleaseClient.cs
using System.Net.Http.Headers;
using System.Security.Cryptography;

namespace ClipSync.Core.Update;

/// <summary>
/// Fetches <c>/repos/{owner}/{repo}/releases/latest</c> and downloads a named
/// asset. The client never applies an update — that is a host concern (WPF
/// restart vs Android package installer).
/// </summary>
public sealed class GitHubReleaseClient : IDisposable
{
    public const string DefaultOwner = "fraternidaddeng";
    public const string DefaultRepo = "clipsync";
    public const string LatestPath = "/releases/latest";

    private readonly HttpClient http;
    private readonly bool ownsHttp;
    private readonly Uri latestUri;

    public GitHubReleaseClient(
        string? currentVersion = null,
        HttpMessageHandler? handler = null,
        Uri? latestUri = null)
    {
        this.latestUri = latestUri ?? new Uri(
            $"https://api.github.com/repos/{DefaultOwner}/{DefaultRepo}{LatestPath}");
        if (handler is null)
        {
            http = new HttpClient(new SocketsHttpHandler
            {
                ConnectTimeout = TimeSpan.FromSeconds(12),
            })
            {
                Timeout = TimeSpan.FromMinutes(15),
            };
            ownsHttp = true;
        }
        else
        {
            http = new HttpClient(handler, disposeHandler: false) { Timeout = TimeSpan.FromMinutes(15) };
            ownsHttp = true;
        }

        var product = string.IsNullOrWhiteSpace(currentVersion) ? "ClipSync" : $"ClipSync/{currentVersion}";
        http.DefaultRequestHeaders.UserAgent.ParseAdd($"{product} (+https://github.com/{DefaultOwner}/{DefaultRepo})");
        http.DefaultRequestHeaders.Accept.Add(new MediaTypeWithQualityHeaderValue("application/vnd.github+json"));
    }

    public async Task<GitHubLatestRelease> FetchLatestAsync(CancellationToken cancellationToken = default)
    {
        var body = await GetTextAsync(latestUri.ToString(), cancellationToken).ConfigureAwait(false);
        return GitHubReleaseParser.Parse(body);
    }

    /// <summary>
    /// Resolves the SHA-256 for <paramref name="payload"/>: prefer the API
    /// <c>digest</c>, otherwise download the <c>.sha256</c> sidecar. Missing
    /// both is a hard failure — the host must not install an unverified file.
    /// </summary>
    public async Task<string> ResolveSha256Async(
        GitHubLatestRelease release,
        ReleaseAsset payload,
        CancellationToken cancellationToken = default)
    {
        if (!string.IsNullOrWhiteSpace(payload.Sha256Hex)
            && GitHubReleaseParser.IsSha256Hex(payload.Sha256Hex))
        {
            return payload.Sha256Hex;
        }

        var sidecar = release.FindSidecar(payload)
            ?? throw new InvalidOperationException(
                $"Release {release.TagName} has no SHA-256 for '{payload.Name}'.");
        var text = await GetTextAsync(sidecar.BrowserDownloadUrl, cancellationToken).ConfigureAwait(false);
        return GitHubReleaseParser.ParseSha256Sidecar(text)
            ?? throw new InvalidOperationException($"Could not parse SHA-256 sidecar for '{payload.Name}'.");
    }

    public async Task DownloadAsync(
        ReleaseAsset asset,
        Stream destination,
        IProgress<UpdateDownloadProgress>? progress = null,
        CancellationToken cancellationToken = default)
    {
        await GetSuccessfulAsync(
                asset.BrowserDownloadUrl,
                async (response, token) =>
                {
                    var total = response.Content.Headers.ContentLength ?? asset.SizeBytes;
                    await using var source = await response.Content
                        .ReadAsStreamAsync(token)
                        .ConfigureAwait(false);
                    var buffer = new byte[81_920];
                    long received = 0;
                    int read;
                    while ((read = await source.ReadAsync(buffer.AsMemory(0, buffer.Length), token)
                               .ConfigureAwait(false)) > 0)
                    {
                        await destination.WriteAsync(buffer.AsMemory(0, read), token).ConfigureAwait(false);
                        received += read;
                        progress?.Report(new UpdateDownloadProgress(received, total));
                    }
                },
                cancellationToken)
            .ConfigureAwait(false);
    }

    private async Task<string> GetTextAsync(string url, CancellationToken cancellationToken)
    {
        string? body = null;
        await GetSuccessfulAsync(
                url,
                async (response, token) =>
                {
                    body = await response.Content.ReadAsStringAsync(token).ConfigureAwait(false);
                },
                cancellationToken)
            .ConfigureAwait(false);
        return body ?? "";
    }

    private async Task GetSuccessfulAsync(
        string url,
        Func<HttpResponseMessage, CancellationToken, Task> handle,
        CancellationToken cancellationToken)
    {
        Exception? lastError = null;
        foreach (var candidate in GitHubUrlMirrors.Candidates(url))
        {
            try
            {
                using var response = await http.GetAsync(
                        new Uri(candidate),
                        HttpCompletionOption.ResponseHeadersRead,
                        cancellationToken)
                    .ConfigureAwait(false);
                if (!response.IsSuccessStatusCode)
                {
                    var errorBody = await response.Content.ReadAsStringAsync(cancellationToken)
                        .ConfigureAwait(false);
                    throw new HttpRequestException(
                        $"GitHub request returned {(int)response.StatusCode}: {TrimForError(errorBody)}");
                }

                await handle(response, cancellationToken).ConfigureAwait(false);
                return;
            }
            catch (Exception exception) when (
                exception is HttpRequestException
                or TaskCanceledException
                or IOException
                or HttpIOException)
            {
                lastError = exception;
            }
        }

        throw lastError ?? new HttpRequestException("No URL candidates.");
    }

    public static string ComputeSha256Hex(Stream stream)
    {
        if (stream.CanSeek)
        {
            stream.Position = 0;
        }

        var hash = SHA256.HashData(stream);
        return Convert.ToHexString(hash).ToLowerInvariant();
    }

    public static void VerifySha256(Stream stream, string expectedHex)
    {
        var actual = ComputeSha256Hex(stream);
        if (!string.Equals(actual, expectedHex, StringComparison.OrdinalIgnoreCase))
        {
            throw new InvalidOperationException(
                $"SHA-256 mismatch (expected {expectedHex}, got {actual}).");
        }
    }

    public void Dispose()
    {
        if (ownsHttp)
        {
            http.Dispose();
        }
    }

    private static string TrimForError(string body)
    {
        var trimmed = body.Trim();
        return trimmed.Length <= 180 ? trimmed : trimmed[..180] + "…";
    }
}

public readonly record struct UpdateDownloadProgress(long ReceivedBytes, long TotalBytes)
{
    public int Percent =>
        TotalBytes > 0 ? (int)Math.Clamp(ReceivedBytes * 100 / TotalBytes, 0, 100) : 0;
}

/// <summary>The host-facing decision after comparing local version to latest.</summary>
public sealed record UpdateCheckResult(
    string CurrentVersion,
    GitHubLatestRelease Latest,
    ReleaseAsset? Payload,
    bool UpdateAvailable)
{
    public static UpdateCheckResult From(
        string currentVersion,
        GitHubLatestRelease latest,
        UpdatePlatform platform)
    {
        var payload = latest.FindPayload(platform);
        var available = payload is not null
            && AppVersion.Compare(currentVersion, latest.VersionLabel) < 0;
        return new UpdateCheckResult(currentVersion, latest, payload, available);
    }
}

File: android/app/src/main/java/com/clipsync/android/update/GitHubReleaseClient.kt
package com.clipsync.android.update

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import java.io.File
import java.io.IOException
import java.security.MessageDigest
import java.util.concurrent.TimeUnit
import kotlin.coroutines.CoroutineContext

/**
 * Fetches `/repos/{owner}/{repo}/releases/latest` and downloads a named asset.
 * Applying the APK is a host concern (PackageInstaller / FileProvider).
 */
class GitHubReleaseClient(
    private val currentVersion: String = "0.0.0",
    private val latestUrl: String = DEFAULT_LATEST_URL,
    private val http: OkHttpClient = defaultClient(),
    private val ioContext: CoroutineContext = Dispatchers.IO,
) {
    suspend fun fetchLatest(): GitHubLatestRelease =
        withContext(ioContext) {
            getText(latestUrl, acceptGithubJson = true) { body ->
                GitHubReleaseParser.parse(body)
            }
        }

    /**
     * Prefer the API `digest`; otherwise download the `.sha256` sidecar. Missing
     * both is a hard failure — the host must not install an unverified file.
     */
    suspend fun resolveSha256(
        release: GitHubLatestRelease,
        payload: ReleaseAsset,
    ): String =
        withContext(ioContext) {
            if (GitHubReleaseParser.isSha256Hex(payload.sha256Hex)) {
                return@withContext payload.sha256Hex!!
            }
            val sidecar =
                release.findSidecar(payload)
                    ?: throw IOException("Release ${release.tagName} has no SHA-256 for '${payload.name}'.")
            getText(sidecar.browserDownloadUrl, acceptGithubJson = false) { body ->
                GitHubReleaseParser.parseSha256Sidecar(body)
                    ?: throw IOException("Could not parse SHA-256 sidecar for '${payload.name}'.")
            }
        }

    suspend fun download(
        asset: ReleaseAsset,
        destination: File,
        onProgress: (received: Long, total: Long) -> Unit = { _, _ -> },
    ) = withContext(ioContext) {
        getSuccessful(asset.browserDownloadUrl, acceptGithubJson = false) { response ->
            val body = response.body ?: throw IOException("Download had no body.")
            val total = body.contentLength().takeIf { it > 0 } ?: asset.sizeBytes
            destination.parentFile?.mkdirs()
            if (destination.exists()) {
                destination.delete()
            }
            destination.outputStream().use { output ->
                body.byteStream().use { input ->
                    val buffer = ByteArray(COPY_BUFFER_BYTES)
                    var received = 0L
                    while (true) {
                        val read = input.read(buffer)
                        if (read < 0) {
                            break
                        }
                        output.write(buffer, 0, read)
                        received += read
                        onProgress(received, total)
                    }
                }
            }
        }
    }

    private fun <T> getText(
        url: String,
        acceptGithubJson: Boolean,
        parse: (String) -> T,
    ): T =
        getSuccessful(url, acceptGithubJson) { response ->
            parse(response.body?.string().orEmpty())
        }

    private fun <T> getSuccessful(
        url: String,
        acceptGithubJson: Boolean,
        handle: (Response) -> T,
    ): T {
        var lastError: IOException? = null
        for (candidate in GitHubUrlMirrors.candidates(url)) {
            try {
                return executeCandidate(candidate, acceptGithubJson, handle)
            } catch (exception: IOException) {
                lastError = exception
            }
        }
        throw lastError ?: IOException("No URL candidates.")
    }

    private fun <T> executeCandidate(
        candidate: String,
        acceptGithubJson: Boolean,
        handle: (Response) -> T,
    ): T {
        val builder =
            Request
                .Builder()
                .url(candidate)
                .header("User-Agent", userAgent(currentVersion))
                .get()
        if (acceptGithubJson) {
            builder.header("Accept", "application/vnd.github+json")
        }
        return http.newCall(builder.build()).execute().use { response ->
            if (!response.isSuccessful) {
                val errorBody = response.body?.string().orEmpty()
                throw IOException(
                    "GitHub request returned ${response.code}: ${errorBody.trim().take(ERROR_BODY_CHARS)}",
                )
            }
            handle(response)
        }
    }

    companion object {
        const val DEFAULT_OWNER = "fraternidaddeng"
        const val DEFAULT_REPO = "clipsync"
        const val DEFAULT_LATEST_URL =
            "https://api.github.com/repos/$DEFAULT_OWNER/$DEFAULT_REPO/releases/latest"
        private const val COPY_BUFFER_BYTES = 81_920
        private const val ERROR_BODY_CHARS = 180
        private const val CONNECT_TIMEOUT_SECONDS = 12L
        private const val TRANSFER_TIMEOUT_MINUTES = 15L
        private const val BYTE_MASK = 0xFF
        private const val NIBBLE_MASK = 0x0F
        private const val NIBBLE_BITS = 4
        private val HEX_DIGITS = "0123456789abcdef".toCharArray()

        fun userAgent(version: String): String = "ClipSync/$version (+https://github.com/$DEFAULT_OWNER/$DEFAULT_REPO)"

        fun defaultClient(): OkHttpClient =
            OkHttpClient
                .Builder()
                .connectTimeout(CONNECT_TIMEOUT_SECONDS, TimeUnit.SECONDS)
                .readTimeout(TRANSFER_TIMEOUT_MINUTES, TimeUnit.MINUTES)
                .writeTimeout(TRANSFER_TIMEOUT_MINUTES, TimeUnit.MINUTES)
                .build()

        fun sha256Hex(file: File): String {
            val digest = MessageDigest.getInstance("SHA-256")
            file.inputStream().use { input ->
                val buffer = ByteArray(COPY_BUFFER_BYTES)
                while (true) {
                    val read = input.read(buffer)
                    if (read < 0) {
                        break
                    }
                    digest.update(buffer, 0, read)
                }
            }
            return toHexLower(digest.digest())
        }

        /**
         * Lowercase hex without going through `String.format("%02x", byte)`.
         * A signed Kotlin/Java `Byte` of `0x80` would otherwise become
         * `ffffff80` and every real APK hash would fail verification.
         */
        internal fun toHexLower(bytes: ByteArray): String {
            val hex = CharArray(bytes.size * 2)
            var index = 0
            for (byte in bytes) {
                val value = byte.toInt() and BYTE_MASK
                hex[index++] = HEX_DIGITS[value ushr NIBBLE_BITS]
                hex[index++] = HEX_DIGITS[value and NIBBLE_MASK]
            }
            return String(hex)
        }

        fun verifySha256(
            file: File,
            expectedHex: String,
        ) {
            val actual = sha256Hex(file)
            if (!actual.equals(expectedHex, ignoreCase = true)) {
                file.delete()
                throw IOException("SHA-256 mismatch (expected $expectedHex, got $actual).")
            }
        }
    }
}
